agent/proxycfg/connect_proxy.go

```diff
@@ -461,7 +462,14 @@ func (s *handlerConnectProxy) handleUpdate(ctx context.Context, u UpdateEvent, s
 				return fmt.Errorf("failed to watch discovery chain for %s: %v", uid, err)
 			}
 		}
-		snap.ConnectProxy.IntentionUpstreams = seenUpstreams
+		snap.ConnectProxy.IntentionUpstreams = intentionUpstreams
+
+		for uid, upstream := range snap.ConnectProxy.UpstreamConfig {
+			// Prevent cleanup of discovery chains for explicit upstream definitions.
+			if !upstream.CentrallyConfigured {
+				seenUpstreams[uid] = struct{}{}
+			}
+		}
 
 		// Clean up data from services that were not in the update
 		for uid, targets := range snap.ConnectProxy.WatchedUpstreams {
```

Comment:
What if instead we modify the check for cross-datacenter upstreams to instead check `!CentrallyConfigured`? That way we avoid allocating a new map. We're already checking UpstreamConfig in the reconciliation. 

```go
if upstream, ok := snap.ConnectProxy.UpstreamConfig[uid]; ok && upstream.Datacenter != "" && upstream.Datacenter != s.source.Datacenter {
	continue
}
```

Could become:
```go
if upstream, ok := snap.ConnectProxy.UpstreamConfig[uid]; ok && !upstream.CentrallyConfigured {
	continue
}
```

This works because a cross-DC watch can only be done explicitly, and not via `service-defaults`/`CentrallyConfigured`.

Reply:
(Applies through L549))